pantos/servicenode/database/access.py

```diff
@@ -471,54 +454,64 @@ def update_transfer_task_id(internal_transfer_id: int,
-
-    """
-    with lock:
-        with get_session_maker().begin() as session:
-            # New session necessary to allow committing after the new
-            # model instance has been added (flush is not sufficient in
-            # a multithreaded environment)
-            instance = session.query(model).filter_by(**kwargs).one_or_none()
-            if instance is None:
-                # Instance has been added by another thread in between
-                instance = model(**kwargs)
-                session.add(instance)
-                session.flush()
-            return instance.id
+def _read_id(session: sqlalchemy.orm.Session, model: typing.Type[B],
+             **kwargs: typing.Any) -> int | None:
+    statement = sqlalchemy.select(model.id).filter_by(**kwargs)
+    return session.execute(statement).scalar_one_or_none()
 
 
-def _create_token_contract(**kwargs: typing.Any) -> int:
-    return _create_instance(TokenContract, _token_contract_lock, **kwargs)
+def _read_forwarder_contract_id(session: sqlalchemy.orm.Session,
+                                blockchain: Blockchain,
+                                address: BlockchainAddress) -> int | None:
+    return _read_id(session, ForwarderContract, blockchain_id=blockchain.value,
+                    address=address)
 
 
-def _read_forwarder_contract(session: sqlalchemy.orm.Session,
-                             **kwargs: typing.Any) -> ForwarderContract:
-    return _read_instance(session, ForwarderContract, **kwargs)
+def _read_hub_contract_id(session: sqlalchemy.orm.Session,
+                          blockchain: Blockchain,
+                          address: BlockchainAddress) -> int | None:
+    return _read_id(session, HubContract, blockchain_id=blockchain.value,
+                    address=address)
 
 
-def _read_hub_contract(session: sqlalchemy.orm.Session,
-                       **kwargs: typing.Any) -> HubContract:
-    return _read_instance(session, HubContract, **kwargs)
+def _read_token_contract_id(session: sqlalchemy.orm.Session,
+                            blockchain: Blockchain,
+                            address: BlockchainAddress) -> int | None:
+    return _read_id(session, TokenContract, blockchain_id=blockchain.value,
+                    address=address)
 
 
-def _read_instance(session: sqlalchemy.orm.Session, model: typing.Type[B],
-                   **kwargs: typing.Any) -> B:
-    return typing.cast(B,
-                       session.query(model).filter_by(**kwargs).one_or_none())
-
-
-def _read_token_contract(session: sqlalchemy.orm.Session,
-                         **kwargs: typing.Any) -> TokenContract:
-    return _read_instance(session, TokenContract, **kwargs)
+def _create_with_id(session: sqlalchemy.orm.Session, model: typing.Type[B],
+                    **kwargs: typing.Any) -> int:
+    statement = sqlalchemy.insert(model).values(**kwargs).returning(model.id)
+    try:
+        with session.begin_nested():
```

Comment:
Neat solution. But just out of curiosity, do you think this would have also worked like this (instead of using a nested transaction)? Or is this necessary? 
```
session.flush()
return session.execute(statement).scalar_one()
```

Reply:
Just for reference: the nested transaction is necessary to prevent the whole database transaction from being rolled back in case of a (perfectly fine) race condition for creating these instances.